<a href="https://colab.research.google.com/github/23f2004341/MachineLearningSeries.Ai-Ml/blob/main/project2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd

df = pd.read_csv('/content/Week3_GA_dataset.csv')
display(df.head())
display(df.info())

V1    V2       V3    V4        V5 Target
0  2.0  50.0  12500.0  98.0  NEGATIVE    YES
1  0.0  13.0   3250.0  28.0  NEGATIVE    YES
2    ?     ?   4000.0  35.0  NEGATIVE    YES
3    ?  20.0   5000.0  45.0  NEGATIVE    YES
4  1.0  24.0   6000.0  77.0  NEGATIVE     NO

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 748 entries, 0 to 747
Data columns (total 6 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   V1      748 non-null    object 
 1   V2      748 non-null    object 
 2   V3      748 non-null    float64
 3   V4      748 non-null    float64
 4   V5      748 non-null    object 
 5   Target  748 non-null    object 
dtypes: float64(2), object(4)
memory usage: 35.2+ KB


None

In [2]:
import numpy as np

df['V1'] = df['V1'].replace('?', np.nan)
df['V2'] = df['V2'].replace('?', np.nan)

df['V1'] = pd.to_numeric(df['V1'])
df['V2'] = pd.to_numeric(df['V2'])

v1_median = df['V1'].median()
v2_median = df['V2'].median()

df['V1'].fillna(v1_median, inplace=True)
df['V2'].fillna(v2_median, inplace=True)

display(df.head())
display(df.info())

/tmp/ipython-input-2858912907.py:12: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['V1'].fillna(v1_median, inplace=True)
/tmp/ipython-input-2858912907.py:13: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'd

V1    V2       V3    V4        V5 Target
0  2.0  50.0  12500.0  98.0  NEGATIVE    YES
1  0.0  13.0   3250.0  28.0  NEGATIVE    YES
2  8.0   4.0   4000.0  35.0  NEGATIVE    YES
3  8.0  20.0   5000.0  45.0  NEGATIVE    YES
4  1.0  24.0   6000.0  77.0  NEGATIVE     NO

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 748 entries, 0 to 747
Data columns (total 6 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   V1      748 non-null    float64
 1   V2      748 non-null    float64
 2   V3      748 non-null    float64
 3   V4      748 non-null    float64
 4   V5      748 non-null    object 
 5   Target  748 non-null    object 
dtypes: float64(4), object(2)
memory usage: 35.2+ KB


None

In [3]:
from sklearn.preprocessing import OneHotEncoder

encoder = OneHotEncoder(handle_unknown='ignore', sparse_output=False)

v5_encoded = encoder.fit_transform(df['V5'].values.reshape(-1, 1))

v5_encoded_df = pd.DataFrame(v5_encoded, columns=encoder.get_feature_names_out(['V5']))

df = pd.concat([df.drop('V5', axis=1), v5_encoded_df], axis=1)

display(df.head())

V1    V2       V3    V4 Target  V5_NEGATIVE
0  2.0  50.0  12500.0  98.0    YES          1.0
1  0.0  13.0   3250.0  28.0    YES          1.0
2  8.0   4.0   4000.0  35.0    YES          1.0
3  8.0  20.0   5000.0  45.0    YES          1.0
4  1.0  24.0   6000.0  77.0     NO          1.0

In [4]:
from sklearn.preprocessing import LabelEncoder

label_encoder = LabelEncoder()
df['Target'] = label_encoder.fit_transform(df['Target'])

display(df.head())

V1    V2       V3    V4  Target  V5_NEGATIVE
0  2.0  50.0  12500.0  98.0       1          1.0
1  0.0  13.0   3250.0  28.0       1          1.0
2  8.0   4.0   4000.0  35.0       1          1.0
3  8.0  20.0   5000.0  45.0       1          1.0
4  1.0  24.0   6000.0  77.0       0          1.0

In [5]:
from sklearn.model_selection import train_test_split

X = df.drop('Target', axis=1)
y = df['Target']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

display(X_train.head())
display(X_test.head())
display(y_train.head())
display(y_test.head())

V1    V2      V3    V4  V5_NEGATIVE
593  9.0  11.0  2750.0  49.0          1.0
131  2.0  12.0  3000.0  95.0          1.0
44   4.0  20.0  5000.0  69.0          1.0
70   2.0   4.0  1000.0  16.0          1.0
670  2.0   3.0   750.0  77.0          1.0

V1   V2      V3    V4  V5_NEGATIVE
580   2.0  1.0   250.0   2.0          1.0
356  16.0  6.0  1500.0  40.0          1.0
133   4.0  6.0  1500.0  35.0          1.0
250  11.0  2.0   500.0  11.0          1.0
299  14.0  2.0   500.0  14.0          1.0

593    1
131    0
44     1
70     0
670    0
Name: Target, dtype: int64

580    0
356    0
133    0
250    0
299    0
Name: Target, dtype: int64

In [6]:
from sklearn.linear_model import LogisticRegression

model = LogisticRegression(random_state=42)
model.fit(X_train, y_train)

LogisticRegression(random_state=42)

In [7]:
from sklearn.metrics import accuracy_score, classification_report

y_pred = model.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy:.4f}")

report = classification_report(y_test, y_pred)
print("Classification Report:")
print(report)

Accuracy: 0.7667
Classification Report:
              precision    recall  f1-score   support

           0       0.77      0.98      0.86       113
           1       0.67      0.11      0.19        37

    accuracy                           0.77       150
   macro avg       0.72      0.55      0.52       150
weighted avg       0.75      0.77      0.70       150



############Summary:

### Data Analysis Key Findings

*   Missing values in columns 'V1' and 'V2' (initially represented as '?') were imputed using the median value of each column.
*   The categorical 'V5' column was one-hot encoded. After preprocessing, it resulted in a single encoded column 'V5\_NEGATIVE'.
*   The 'Target' variable was successfully encoded into numerical labels (0 and 1).
*   The dataset was split into training and testing sets with 80% for training and 20% for testing.
*   A Logistic Regression model was trained on the training data.
*   The trained model achieved an overall accuracy of approximately 76.67% on the test set.
*   The model demonstrated high precision (0.77) and recall (0.98) for class 0 but showed lower precision (0.67) and significantly low recall (0.11) for class 1, indicating difficulty in identifying instances of class 1.


